Implement translation between External Metrics API and Stackdriver API

kawych opened the pull request:
This PR adds a code for translation between External Metrics API and Stackdriver API. To use this code, Custom Metrics - Stackdriver Adapter will have to expose External Metrics API endpoint, similarly to this PR: https://github.com/kubernetes-incubator/custom-metrics-apiserver/pull/15

kawych commented:
@MaciekPytel PTAL

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/provider.go:
```diff
@@ -201,6 +203,30 @@ func (p *StackdriverProvider) ListAllMetrics() []provider.MetricInfo {
+}
+
+func unescapeMetricName(metricName string) string {
+	return strings.Replace(metricName, "|\\", "/", -1)
```
Can we change this to just a pipe ('|')? I wanted a longer escape sequence when it was meant for multiple different systems, but AFAIK pipe is not allowed in sd names, so it's good enough by itself.

kawych replied:
Done

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:
```diff
@@ -137,6 +161,44 @@ func (t *Translator) GetRespForMultipleObjects(response *stackdriver.ListTimeSer
+	metricValues := []external_metrics.ExternalMetricValue{}
+	// Find time series with specified labels matching
+	// Stackdriver API doesn't allow complex label filtering (i.e. "label1 = x AND (label2 = y OR label2 = z)"),
+	// therefore only part of the filters is passed and remaining filtering is done here.
```
This comments says there should be some filtering here, but I see nothing of the sort. Am I missing something? Or is there some code missing here?

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:
```diff
@@ -187,6 +249,29 @@ func getNodeNames(list *v1.NodeList) []string {
+			return prefix, strings.TrimPrefix(labelName, prefix+"."), nil
+		}
+	}
+	return "", "", fmt.Errorf("TODO")
```
??

kawych replied:
Done

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:
```diff
@@ -265,6 +358,87 @@ func (t *Translator) legacyFilterForPods(podIDs []string) string {
+func (t *Translator) filterForSelector(metricSelector labels.Selector) (string, error) {
+	requirements, selectable := metricSelector.Requirements()
+	if !selectable {
+		return "", provider.NewOperationNotSupportedError("TODO")
```
Add error message / error handling

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator.go:
```diff
@@ -265,6 +358,87 @@ func (t *Translator) legacyFilterForPods(podIDs []string) string {
+	for label, value := range series.Resource.Labels {
+		metricLabels["resource.label."+label] = value
+	}
+	return metricLabels
```
Should we include metadata labels here? I don't know if we should, but we do allow them in query.

kawych replied:
I think we should not. Metric and resource labels are exported with each metric. Metadata labels are exported by different channel to provide more information about given resource to Stackdriver. But resource labels include a complete set of information required to identify the resource.

For example, pods are identified by resource labels: cluster name, location, project id, namespace and name. Metadata contain information such as node name, pod labels, etc.

MaciekPytel replied:
So you can use metadata labels to query based on pod labels, and every metric value will come with a set of labels identifying it (namespace + name), correct? Makes sense.

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator_test.go:
```diff
@@ -369,6 +372,97 @@ func TestTranslator_GetMetricsFromSDDescriptorsResp(t *testing.T) {
+			"AND metric.label.my_label > 86 " +
+			"AND NOT resource.label.namespace_name = one_of(\"default\",\"kube-system\") " +
+			"AND resource.label : pod_name " +
+			"AND resource.type = \"k8s_pod\"").
```
Is the order of conditions well defined or does the ordering in expected query just randomly work?

kawych replied:
They're sorted by key: https://github.com/kubernetes/apimachinery/blob/master/pkg/labels/selector.go#L100

MaciekPytel replied:
Cool, that answers my question.

MaciekPytel commented on custom-metrics-stackdriver-adapter/pkg/adapter/provider/translator_test.go:
```diff
@@ -369,6 +372,97 @@ func TestTranslator_GetMetricsFromSDDescriptorsResp(t *testing.T) {
+	}
+}
+
+func TestTranslator_GetExternalMetricRequest(t *testing.T) {
```
It feels like the test name is missing suffix

kawych replied:
Done

MaciekPytel commented:
I think we could use a few more UT covering more filters (ex. a combination of few filters where only one is incorrect, or a query with more than one condition of given type (label1 = value1 and label2 = value2)).

kawych commented:
Added a test for more filters where one is incorrect. I've added the other case to one of existing ones, although I think it doesn't do much, since all requirements are translated separately.

MaciekPytel commented:
/lgtm
